code/harry_potter_store/src/store.py

```diff
@@ -1,33 +1,75 @@
+from copy import deepcopy
+from typing import Dict
+
+
 class Store:
     book_types = ["1", "2", "3", "4", "5"]
 
     def __init__(self, book_price=8):
         self.book_price = book_price
         pass
 
-    def _clean(self, to_buy_books: dict) -> dict:
+    def _validate(self, to_buy_books: Dict[str, int]):
+        for book_type, amount in to_buy_books.items():
+            if book_type not in Store.book_types:
+                raise ValueError(f"Book type {book_type} must be in {Store.book_types}")
+            if amount <= 0:
+                raise ValueError(f"Book amount {amount} for type {book_type} must be > 0")
+
+    def _clean(self, to_buy_books: Dict[str, int]) -> dict:
         return dict(filter(lambda item: item[1] > 0 and item[0] in Store.book_types, to_buy_books.items()))
 
-    def _decrement(self, to_buy_books: dict) -> dict:
-        return self._clean(dict(map(lambda i: [i[0], i[1] - 1], to_buy_books.items())))
+    def _decrement(self, to_buy_books: Dict[str, int], count: int) -> dict:
+        first_part = {book: amount - 1 for book, amount in list(to_buy_books.items())[:count]}
+        second_part = dict(list(to_buy_books.items())[count:])
+        first_part.update(second_part)
+        return self._clean(first_part)
+
+    def _discount_2(self, to_buy_books, price, predicate):
+        if predicate(len(to_buy_books), 2):
+            price += 2 * self.book_price * 0.95
+            to_buy_books = self._decrement(to_buy_books, 2)
+        return to_buy_books, price
+
+    def _discount_3(self, to_buy_books, price, predicate):
+        if predicate(len(to_buy_books), 3):
+            price += 3 * self.book_price * 0.90
+            to_buy_books = self._decrement(to_buy_books, 3)
+        return to_buy_books, price
+
+    def _discount_4(self, to_buy_books, price, predicate):
+        if predicate(len(to_buy_books), 4):
+            price += 4 * self.book_price * 0.80
+            to_buy_books = self._decrement(to_buy_books, 4)
+        return to_buy_books, price
+
+    def _discount_5(self, to_buy_books, price, predicate):
+        if predicate(len(to_buy_books), 5):
+            price += 5 * self.book_price * 0.75
+            to_buy_books = self._decrement(to_buy_books, 5)
+        return to_buy_books, price
+
+    """
+    to_buy_books - dictionary from book type and book amount
+    key of dict -- book type
+    value of dict -- book amount  
+    """
+
+    def get_price(self, to_buy_books: Dict[str, int]) -> float:
+        self._validate(to_buy_books)
+        discounts = [self._discount_5, self._discount_4, self._discount_3, self._discount_2]
+        prices = []
+        for i in range(len(discounts)):
```

Comment:
I don't understand what you want.

Reply:
Make the algorithm simpler :) Remove extra iterations and not needed parameters, add some comments explaining the algorithm and improving readability.

Reply:
Can you describe way to simplification. I already was doing it simply and you rejected it.

Reply:
No-no, I didn't "reject" anything, I suggested a few options and you have selected one =) If you think one or another review suggestion doesn't make sense or wrong or something was done intentionally - you can always say it, code review is a dual-way process :)

>Can you describe way to simplification.
I have described it in the top-level comment, didn't I?

Simplify the forloops:
>I think the second case now is too safe / doing too much =) Cannot it be replaced to just one check with the whole discounts array? Like self._get_price(to_buy_books, discounts, 1, lambda x, y: x == y)

Remove extra parameters from the functions:
>Then, if this is done, it can be refactored further to remove min_value parameter as it's not really needed IMO (while condition can be checked i.e. as "did the price change after full iteration?")

Add comments to not obvious places and parameters to make it easier for reader to follow your algorithm:

>P.S. It would be still great to have some comments in the code explaining the algorithm...

Reply:
fix sorting in algorithm and add more docs